src/main/java/org/zalando/nakadi/service/CursorsService.java

```diff
@@ -50,43 +51,52 @@
     private final ZooKeeperLockFactory zkLockFactory;
     private final ZkSubscriptionClientFactory zkSubscriptionClientFactory;
     private final SubscriptionKafkaClientFactory subscriptionKafkaClientFactory;
+    private final CursorTokenService cursorTokenService;
 
     @Autowired
     public CursorsService(final ZooKeeperHolder zkHolder,
-                                final TopicRepository topicRepository,
-                                final SubscriptionDbRepository subscriptionRepository,
-                                final EventTypeRepository eventTypeRepository,
-                                final ZooKeeperLockFactory zkLockFactory,
-                                final ZkSubscriptionClientFactory zkSubscriptionClientFactory,
-                                final SubscriptionKafkaClientFactory subscriptionKafkaClientFactory) {
+                          final TopicRepository topicRepository,
+                          final SubscriptionDbRepository subscriptionRepository,
+                          final EventTypeRepository eventTypeRepository,
+                          final ZooKeeperLockFactory zkLockFactory,
+                          final ZkSubscriptionClientFactory zkSubscriptionClientFactory,
+                          final SubscriptionKafkaClientFactory subscriptionKafkaClientFactory,
+                          final CursorTokenService cursorTokenService) {
         this.zkHolder = zkHolder;
         this.topicRepository = topicRepository;
         this.subscriptionRepository = subscriptionRepository;
         this.eventTypeRepository = eventTypeRepository;
         this.zkLockFactory = zkLockFactory;
         this.zkSubscriptionClientFactory = zkSubscriptionClientFactory;
         this.subscriptionKafkaClientFactory = subscriptionKafkaClientFactory;
+        this.cursorTokenService = cursorTokenService;
     }
 
-    public boolean commitCursors(final String subscriptionId, final List<Cursor> cursors)
+    public boolean commitCursors(final String subscriptionId, final List<SubscriptionCursor> cursors)
             throws NakadiException, InvalidCursorException {
 
         final Subscription subscription = subscriptionRepository.getSubscription(subscriptionId);
-        final String eventTypeName = subscription.getEventTypes().iterator().next();
-        final EventType eventType = eventTypeRepository.findByName(eventTypeName);
-
         createSubscriptionInZkIfNeeded(subscription);
-        topicRepository.validateCommitCursors(eventType.getTopic(), cursors);
 
         boolean allCommitted = true;
-        for (final Cursor cursor : cursors) {
-            final boolean cursorCommitted = commitCursor(subscriptionId, eventType.getTopic(), cursor);
-            allCommitted = allCommitted && cursorCommitted;
+        for (final String eventTypeName : subscription.getEventTypes()) {
```

Comment:
Yes, it really makes sense to iterate by cursors. I changed that, now the code is more clear.